src/sentry/api/endpoints/artifact_lookup.py

```diff
@@ -0,0 +1,336 @@
+    dist = None
+    # TODO: Is there a way to safely query this and return `None` if not existing?
+    try:
+        release = Release.objects.get(
+            organization_id=project.organization_id,
+            projects=project,
+            version=release_name,
+        )
+
+        # We cannot query for dist without a release anyway
+        if dist_name:
+            dist = Distribution.objects.get(release=release, name=dist_name)
+    except Exception as exc:
+        logger.error("Failed to read", exc_info=exc)
+
+    urls = collect_legacy_artifact_bundles_containing_urls(urls, release, dist, bundle_file_ids)
+    if not urls:
+        return list()
+
+    # And last but not least, we want to look up legacy individual release files
+    return get_releasefiles_matching_urls(urls, release)
+
+
+def collect_release_artifact_bundles_containing_urls(
+    urls: List[str], project: Project, release_name: str, dist_name: str, bundle_file_ids: Set[int]
+) -> List[str]:
+    releases_with_bundles = ReleaseArtifactBundle.objects.filter(
+        organization_id=project.organization.id,
+        release_name=release_name,
+        dist_name=dist_name,
+    ).select_related("artifact_bundle__file")[:MAX_SCANNED_BUNDLES]
+
+    manifests = []
+    for release in releases_with_bundles:
+        file_id = release.artifact_bundle.file.id
+        with release.artifact_bundle.file.getfile() as file:
+            archive = ArtifactBundleArchive(file)
+            manifest = archive._read_manifest()
+            manifests.append((file_id, manifest))
+
+    def url_in_any_manifest(url):
+        for (file_id, manifest) in manifests:
+            if url_exists_in_manifest(manifest, url):
+                bundle_file_ids.add(file_id)
+                return True
+        return False
+
+    return list(filter(lambda url: not url_in_any_manifest(url), urls))
+
+
+def collect_legacy_artifact_bundles_containing_urls(
+    urls: List[str], release: Release, dist: Optional[Distribution], bundle_file_ids: Set[int]
+) -> List[str]:
+    artifact_index = None
+    try:
+        artifact_index = read_artifact_index(release, dist, artifact_count__gt=0)
+    except Exception as exc:
+        logger.error("Failed to read artifact index", exc_info=exc)
+    if not artifact_index:
+        return urls
+
+    artifact_archives = dict()
+
+    def url_in_any_artifact_index(url):
+        file = find_file_in_archive_index(artifact_index, url)
+        if file is not None:
+            ident = file["archive_ident"]
+            archive = artifact_archives.get(ident)
+
+            if archive is not None:
+                bundle_file_ids.add(archive)
+            else:
+                artifact_index_qs = ReleaseFile.objects.filter(
+                    release_id=release.id, ident=ident
+                ).select_related("file")
+                artifact_archives[ident] = artifact_index_qs[0].file.id
+                bundle_file_ids.add(artifact_index_qs[0].file.id)
```

Comment:
From my limited understanding of query sets, this will result in two identical queries. We could just get `[0]` after the `.select_related("file")`.